### DISCUSSION TITLE
Improving 'mob' with Remote Mob Programming

### DISCUSSION THREAD

**Discussion by simonharrer** [remotemobprogramming/mob#92] [General]
We want to improve the mob tool via Remote Mob Programming. In this conversation, we try to find a date and time for the sessions.

**Comment by simonharrer** [comment]
Goal:
- push https://github.com/remotemobprogramming/mob/issues/78 forward

Availability Simon
- 2020-12-22 20 to 23, CET
- 2020-12-28 20 to 23, CET
- 2020-12-29 20 to 23, CET
- 2020-12-30 20 to 23, CET

Please write your availability in a reply.

**Reply by gregorriegler** [reply]
let me know when you have decided on a date. I would like to attend if I can find the time.

**Reply by jbrains** [reply]
I'm available:

- 2020-12-22
- 2020-12-29

**Reply by koeberlue** [reply]
Availability Markus

    2020-12-28 20 to 23, CET
    2020-12-29 20 to 23, CET
    2020-12-30 20 to 23, CET

Since my daughter was born a week ago (🎉 ) i might have to cancel on short notice and possibly won't make it till 23:00. But i will try my best to be part of the experience 😄 

@simonharrer Should we prepare a local setup? Like a specific IDE or certain collaboration tools?

**Reply by simonharrer** [reply]
Okay, let's try 2020-12-29 20 to 23, CET.

@koeberlue congratulations! Family comes first. My second daughter was born in January, so I have a good idea what you are going through now. If you're there, it's great. And if you have to take care of your child or yourself, that's great too! :-)

Setup:
- have 'mob' installed
- have 'golang' installed
- I use IntelliJ Ultimate but any IDE is ok

I invite everybody to be collaborators of this repo so you can use 'mob' (you need access rights to push to the wip branch).

**Reply by simonharrer** [reply]
https://innoq.zoom.us/j/93799303693 with a waiting room.

**Comment by jbrains** [comment]
I am new to golang. From IDEA I can build and run the application, but I can't run the tests. I can't even run them from the command line. What very simple step am I missing?

```
# I'm in the project root directory
$ go test mob_test.go     
# command-line-arguments [command-line-arguments.test]
./mob_test.go:17:2: undefined: configuration
./mob_test.go:17:18: undefined: getDefaultConfiguration
./mob_test.go:19:12: undefined: configuration
./mob_test.go:20:25: undefined: parseArgs
./mob_test.go:24:21: undefined: configuration
./mob_test.go:28:2: undefined: configuration
./mob_test.go:28:18: undefined: getDefaultConfiguration
./mob_test.go:30:12: undefined: configuration
./mob_test.go:31:25: undefined: parseArgs
./mob_test.go:35:23: undefined: configuration
./mob_test.go:35:23: too many errors
FAIL    command-line-arguments [build failed]
FAIL
````

**Reply by jbrains** [reply]
Aha! I can run the tests from the command line, but they fail miserably, so I don't know what assumptions the tests are making that I'm not.

I get a bunch of this:

```
SET WORKING DIR TO /tmp/mob/local
======================
mob_test.go:583:

        exp: "origin/master"

        got: ""

--- FAIL: TestVersion (0.22s)
    mob_test.go:603: command ./create-testbed failed
    mob_test.go:704: goroutine 11 [running]:
    [...]
```

**Reply by simonharrer** [reply]
Have a look at https://github.com/remotemobprogramming/mob#how-to-contribute

I run the tests with „go test“ only.

**Reply by simonharrer** [reply]
The tests require a testbed. Have a look at the create-testbed bash script in the root dir. Can you run that on its own on your machine?

**Reply by jbrains** [reply]
I'll look into this today and I might propose some improvements to the documentation for new contributors. I will probably find some undocumented assumptions along the way. :)

**Reply by simonharrer** [reply]
Sure. Go ahead. :-)

**Reply by gregorriegler** [reply]
Ok, also preparing my development setup now. Also new to Golang, let's see how this goes :-)

**Reply by jbrains** [reply]
I believe I'm ready to go. I can run tests from both IDEA and the command line. I can also build the application from both places.

IDEA warns me of an empty `GOPATH`. Should I do anything about that? I prefer not to silence the warning forever, but instead fix the problem.

@gregorriegler might find this PR helpful: https://github.com/remotemobprogramming/mob/pull/93

Thanks.

**Reply by gregorriegler** [reply]
thank you @jbrains . i had all the same issues as you, will try your fix tomorrow. i had the GOPATH thing as well. double checked the golang installation guide, but it says nothing about the GOPATH. found this https://golang.org/doc/gopath_code.html#GOPATH but it works without it

**Reply by jbrains** [reply]
It looks like `GOPATH` won't become important until I have multiple projects in golang, which I don't intend to do soon.

**Reply by simonharrer** [reply]
Hm, don't know anymore how I went about the GOPATH variable. In IntelliJ IDEA it looks like this (maybe this helps):

<img width="182" alt="Screenshot 2020-12-23 at 09 17 23" src="https://user-images.githubusercontent.com/375693/102975013-b80ad900-44ff-11eb-9a94-ab59ad72ce49.png">

**Reply by jbrains** [reply]
It looks the "load path" from other languages: where to search for libraries outside the project.

**Reply by gregorriegler** [reply]
@jbrains @simonharrer I found the way to run the tests from within IntelliJ.
Maybe this will help you. Made a pull request.

https://github.com/remotemobprogramming/mob/pull/94/commits/e815f03b0c39b3bba9caa30fecd2459e1843f03e
The tests also create the test bed automatically for me, and the tests just work.
I even tried deleting the test bed first. the tests recreated it, and it worked.
So on my machine the test bed is created automatically - no manual script run required.

**Reply by jbrains** [reply]
@gregorriegler Thanks for that. Now that I can read the code better, I see that the tests explicitly try to run `create-testbed`. I don't know why it didn't work for me when I tried it the first time.

**Reply by gregorriegler** [reply]
Also, I recently figured out that with the adding of the `go.mod` file, IntelliJ somehow detected the "Go module" and now allows me to run Tests with "Kind: Package".
This is even better as it allows me to see **code coverage in the IDE** which it previously didn't. And it allows me to **debug a test in the IDE**, which it previously didn't either.
